File: Day4/day4.py
from itertools import tee
import sys

def intervals_contained(intervals: list[list[int]]) -> bool:
    return (intervals[0][0] - intervals[1][0]) * (
        intervals[0][1] - intervals[1][1]
    ) <= 0


def intervals_overlap(intervals: list[list[int]]) -> bool:
    return max(intervals[0][0], intervals[1][0]) <= min(
        intervals[0][1], intervals[1][1]
    )


def split_sections_into_intervals(string: str) -> list[list[int]]:
    return [
        [int(number) for number in item.split("-")] for item in string.split(",")
    ]

if __name__ == "__main__":
    try:
        file_name = sys.argv[1]
    except IndexError:
        file_name = "input.txt"
    data = open(file_name).read().split(sep="\n")
    data.remove("")
    interval_extremes_1, interval_extremes_2 = tee(
        map(split_sections_into_intervals, data), 2
    )
    score_round_1 = sum(map(intervals_contained, interval_extremes_1))
    score_round_2 = sum(map(intervals_overlap, interval_extremes_2))

    print(f"The number of contained intervals is {score_round_1}")
    print(f"The number of overlaps in round two is {score_round_2}")

File: Day4/day4.rs
use std::cmp;

fn get_data() -> String {
    let input_args: Vec<String> = std::env::args().collect();
    let mut file_name: String = "input.txt".to_string();
    if input_args.len() > 1 {
        file_name = input_args[1].clone();
    }
    let mut data = std::fs::read_to_string(file_name).unwrap();
    if data.ends_with("\n") {
        data.pop();
    }
    return data;
}

fn parse_interval_extremes(input_line: &str) -> Vec<i32> {
    return input_line
        .split(|c| c == ',' || c == '-')
        .map(|element| element.parse::<i32>().unwrap())
        .collect();
}

fn intervals_contained(intervals: Vec<i32>) -> bool {
    return (intervals[0] - intervals[2]) * (intervals[1] - intervals[3]) <= 0;
}

fn intervals_overlap(intervals: Vec<i32>) -> bool {
    return cmp::max(intervals[0], intervals[2]) <= cmp::min(intervals[1], intervals[3]);
}

fn main() {
    let data = get_data();
    let interal_extremes = data.split("\n").map(|item| parse_interval_extremes(item));
    let result_1 = interal_extremes
        .clone()
        .map(|vector| intervals_contained(vector))
        .into_iter()
        .filter(|b| *b)
        .count();
    let result_2 = interal_extremes
        .map(|vector| intervals_overlap(vector))
        .into_iter()
        .filter(|b| *b)
        .count();
    println!("The number of contained intervals is {}", result_1);
    println!("The number of overlaping intervals is {}", result_2);
}
